src/crypto/sodium.cpp

```diff
@@ -22,138 +22,190 @@
+    strcpy(result, fpHex.c_str());
+
+    return result;
+}
+
+void EdDSA::signKey(const unsigned char *key, const unsigned long long keyLength, string *result)
+{
+    uint64_t ts = (uint64_t) time(NULL);
+
+    string tsstr;
+    unsigned char digit;
+    for (int i = 0; i < 8; i++)
     {
-        memcpy(privKey, this->keySeed, crypto_sign_SEEDBYTES);
+        digit = ts & 0xFF;
+        tsstr.insert(0, 1, digit);
+        ts >>= 8;
     }
-    return(1);
+
+    string keyString = "keyauth";
+    keyString.append(tsstr);
+    keyString.append((char*)key, keyLength);
+
+    byte sigBuf[crypto_sign_BYTES];
+    sign((unsigned char *)keyString.data(), keyString.size(), sigBuf);
+
+    result->resize(crypto_sign_BYTES + sizeof(ts));  // 8 --> timestamp prefix
+    result->assign(tsstr.data(), sizeof(ts));
+    result->append((const char*)sigBuf, crypto_sign_BYTES);
 }
 
+bool EdDSA::verifyKey(const unsigned char *pubk, const unsigned long long pubkLen, string *sig, const unsigned char* signingPubKey)
+{
+    uint64_t ts;
+    memcpy(&ts, sig->substr(0, 8).data(), sizeof(ts));
+
+    string message = "keyauth";
+    message.append(sig->data(), 8);
+    message.append((char*)pubk, pubkLen);
 
-// Derives the Ed25519 public key from the stored private key seed.
-int EdDSA::publicKey(unsigned char* pubKey) {
-    unsigned char* privKey = (unsigned char*)malloc(crypto_sign_SECRETKEYBYTES);
-    if (privKey == NULL) {
-        // Something went wrong allocating the memory.
-        return(0);
+    string signature = sig->substr(8);
+
+    return verify((unsigned char*) message.data(), message.length(),
+                  (unsigned char*) signature.data(),
+                  signingPubKey ? signingPubKey : pubKey);
+}
+
+const std::string ECDH::TLV_KEY= "prCu255";
+
+ECDH::ECDH(unsigned char *privKey)
+{
+    if (sodium_init() == -1)
+    {
+        LOG_err << "Cannot initialize sodium library.";
+        return;
     }
-    int check = crypto_sign_seed_keypair(pubKey, privKey,
-                                         (const unsigned char*)this->keySeed);
-    if (check != 0) {
-        // Something went wrong deriving keys.
-        free(privKey);
-        return(0);
+
+    if (privKey)    // then use the value
+    {
+        memcpy(this->privKey, privKey, PRIVATE_KEY_LENGTH);
+
+        // derive public key from privKey
+        crypto_scalarmult_base(this->pubKey, this->privKey);
     }
-    free(privKey);
-    return(1);
+    else
+    {
+        // no private key specified: create a new key pair
+        crypto_box_keypair(this->pubKey, this->privKey);
+    }
+}
+
+ECDH::~ECDH()
+{
+}
+
+int ECDH::encrypt(unsigned char *encmsg, const unsigned char *msg,
+                  const unsigned long long msglen, const unsigned char *nonce,
+                  const unsigned char *pubKey, const unsigned char *privKey)
+{
+    int check = crypto_box(encmsg, msg, msglen, nonce, pubKey, privKey);
+
+    return check ? 0 : 1;
```

Comment:
This could be simplified returning `!check`, or directly returning `!crypto_box...`